clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java

```diff
@@ -236,6 +241,27 @@ public ChannelMuteState muteState() {
         return muteState;
     }
 
+    /**
+     * Mute this channel on {@link #prepare()} failure. This will remove all operations from this channel.
+     */
+    private void muteOnPrepareFailure() {
+        mute();
+        transportLayer.removeInterestOps(SelectionKey.OP_WRITE | SelectionKey.OP_CONNECT | SelectionKey.OP_ACCEPT);
+    }
+
+    /**
+     * Finish up any processing on {@link #prepare()} failure.
+     * @throws IOException
+     */
+    void completeCloseOnPrepareFailure() throws IOException {
+        // If transport layer is ready but authentication failed, invoke the underlying handler to finish up any
+        // processing on authentication failure
+        if (transportLayer.ready()) {
```

Comment:
I think this is working around SSL, but it is an odd check. Shouldn't we just do the operations below anyway? In any case, I am not sure if this approach works for SSL. If we have already written the SSL handshake failure before we delay the connection close, the client would see the handshake failure, close the connection and attempt to reconnect. Holding on the connection in the broker in this case would potentially leave a lot of connections in CLOSE_WAIT state.